README.rst
===========
farneback3d
===========

.. image:: https://badge.fury.io/py/farneback3d.svg
    :target: https://badge.fury.io/py/farneback3d
.. image:: https://travis-ci.org/theHamsta/farnback3d.svg?branch=master
    :target: https://travis-ci.org/theHamsta/farnback3d


A CUDA implementation of the Farneback optical flow algorithm [1]_ for the calculation of dense volumetric flow fields. Since this algorithm is based on the approximation of the signal by polynomial expansion it is especial suited for the motion estimation in smooth signals without clear edges.

To know more about the implementation have a look on `this OpenCV class <https://docs.opencv.org/3.3.0/de/d9e/classcv_1_1FarnebackOpticalFlow.html>`_ that was used as a template for this implementation.

Development takes place on `Github <https://github.com/theHamsta/farneback3d>`_.

Python interface
================

The project uses `pycuda <https://github.com/inducer/pycuda>`_ to provide a pure-python package available on PyPi

.. code-block:: bash

    pip install farneback3d

Usage:

.. code-block:: python

    import farneback3d

    ... # create some numpy volumes vol0 and vol1 (can also be pycuda GPUArrays) 

    # set parameters for optical flow (parameters have the same meaning as in the OpenCV implementation)
    optflow = farneback3d.Farneback(
            pyr_scale=0.8,         # Scaling between multi-scale pyramid levels
            levels=6,              # Number of multi-scale levels
            num_iterations=5,      # Iterations on each multi-scale level
            winsize=9,             # Window size for Gaussian filtering of polynomial coefficients
            poly_n=5,              # Size of window for weighted least-square estimation of polynomial coefficients
            poly_sigma=1.2,        # Sigma for Gaussian weighting of least-square estimation of polynomial coefficients
        )

    # calculate frame-to-frame flow between vol0 and vol1
    flow = optflow.calc_flow(vol0, vol1)


C++ interface
=============

To be implemented...


Future plans
=============

The current implementation uses a naive approach to perform the necessary convolutions.
The algorithm could be sped up drastically by performing separable convolutions along each coordinate axis.

References
==========

.. [1] Farnebäck, Gunnar. "Two-frame motion estimation based on polynomial expansion." Scandinavian conference on Image analysis. Springer, Berlin, Heidelberg, 2003.


CHANGES.rst
=========
Changelog
=========

Version 0.1
===========

- add code used for paper


tests/conftest.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""
    Dummy conftest.py for farneback3d.

    If you don't know what this is for, just leave it empty.
    Read more about conftest.py under:
    https://pytest.org/latest/plugins.html
"""
from __future__ import print_function, absolute_import, division

import pytest


tests/test_skeleton.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import pytest
from farneback3d.skeleton import fib

__author__ = "Stephan Seitz"
__copyright__ = "Stephan Seitz"
__license__ = "none"


def test_fib():
    assert fib(1) == 1
    assert fib(2) == 1
    assert fib(7) == 13
    with pytest.raises(AssertionError):
        fib(-10)


tests/test_synthetic_volumes.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import pytest
import farneback3d
import numpy as np


__author__ = "Stephan Seitz"
__copyright__ = "Stephan Seitz"
__license__ = "none"

TOLERANCE = 1e-2

def test_moving_cube():

    movement_vector = [0.3, 0.1, 1]

    vol0 = np.zeros([60, 60, 60], np.float32)
    vol0[10:15, 20:25, 30:35] = 100.
    # flow_ground_truth = 4* np.ones([*vol0.shape,3], np.float32)
    flow_ground_truth = np.stack([movement_vector[2] * np.ones(vol0.shape, np.float32), movement_vector[1] * np.ones(
        vol0.shape, np.float32), movement_vector[0] * np.ones(vol0.shape, np.float32)], 0)

    vol1 = farneback3d.warp_by_flow(vol0, flow_ground_truth)

    optflow = farneback3d.Farneback(
        levels=5, num_iterations=5, poly_n=5, quit_at_level=-1, use_gpu=True, fast_gpu_scaling=True)
    flow = optflow.calc_flow(vol0, vol1)

    for axis in range(len(movement_vector)):
        print(np.max(flow[2-axis]) - movement_vector[axis])
        assert abs(np.max(flow[2-axis]) - movement_vector[axis]) < TOLERANCE

def test_moving_cube_larger_distance():

    movement_vector = [4, 1, 1]

    vol0 = np.zeros([60, 60, 60], np.float32)
    vol0[10:15, 20:25, 30:35] = 100.
    # flow_ground_truth = 4* np.ones([*vol0.shape,3], np.float32)
    flow_ground_truth = np.stack([movement_vector[2] * np.ones(vol0.shape, np.float32), movement_vector[1] * np.ones(
        vol0.shape, np.float32), movement_vector[0] * np.ones(vol0.shape, np.float32)], 0)

    vol1 = farneback3d.warp_by_flow(vol0, flow_ground_truth)

    optflow = farneback3d.Farneback(
        levels=5, num_iterations=5, poly_n=5, quit_at_level=-1, use_gpu=True, fast_gpu_scaling=True)
    flow = optflow.calc_flow(vol0, vol1)

    for axis in [0]:
        print(np.max(flow[2-axis]) - movement_vector[axis])
        assert abs(np.max(flow[2-axis]) - movement_vector[axis]) < TOLERANCE

if __name__ == "__main__":
    test_moving_cube()
    test_moving_cube_larger_distance()


farneback3d/_utils.py
import os
import numpy as np
import pycuda.autoinit
import pycuda.driver as cuda
from pycuda import gpuarray
from pycuda.compiler import SourceModule


def divup(a, b):
    if a % b:
        return a / b + 1
    else:
        return a / b


def ndarray_to_float_tex(tex_ref, ndarray, address_mode=cuda.address_mode.BORDER, filter_mode=cuda.filter_mode.LINEAR):
    if isinstance(ndarray, np.ndarray):
        cu_array = cuda.np_to_array(ndarray, 'C')
    elif isinstance(ndarray, gpuarray.GPUArray):
        cu_array = cuda.gpuarray_to_array(ndarray, 'C')
    else:
        raise TypeError(
            'ndarray must be numpy.ndarray or pycuda.gpuarray.GPUArray')

    cuda.TextureReference.set_array(tex_ref, cu_array)

    cuda.TextureReference.set_address_mode(
        tex_ref, 0, address_mode)
    if ndarray.ndim >= 2:
        cuda.TextureReference.set_address_mode(
            tex_ref, 1, address_mode)
    if ndarray.ndim >= 3:
        cuda.TextureReference.set_address_mode(
            tex_ref, 2,  address_mode)
    cuda.TextureReference.set_filter_mode(
        tex_ref, filter_mode)
    tex_ref.set_flags(tex_ref.get_flags(
    ) & ~cuda.TRSF_NORMALIZED_COORDINATES & ~cuda.TRSF_READ_AS_INTEGER)


with open(os.path.join(os.path.dirname(__file__), 'resize.cu')) as f:
    _read_data = f.read()

_mod = SourceModule(_read_data)
_tex_ref = _mod.get_texref('sourceTex')
_kernel = _mod.get_function('resize')


def resize_gpu(src_vol, dst_vol=None, dst_shape=None, scaling=None):

    if dst_shape is None:
        assert scaling
        dst_shape = [np.int(np.round(i * scaling)) for i in src_vol.shape]

    if dst_vol is None:
        dst_vol = gpuarray.GPUArray(dst_shape, np.float32)

    ndarray_to_float_tex(_tex_ref, src_vol)

    block = (32, 32, 1)
    grid = (int(divup(dst_vol.shape[2], block[0])),
            int(divup(dst_vol.shape[1], block[1])), 1)

    _kernel(dst_vol,
            np.int32(src_vol.shape[2]),
            np.int32(src_vol.shape[1]),
            np.int32(src_vol.shape[0]),
            np.int32(dst_vol.shape[2]),
            np.int32(dst_vol.shape[1]),
            np.int32(dst_vol.shape[0]),
            grid=grid, block=block)

    return dst_vol


# from https://www.andreas-jung.com/contents/a-python-decorator-for-measuring-the-execution-time-of-methods
def timeit(method):

    def timed(*args, **kw):
        ts = time.time()
        result = method(*args, **kw)
        te = time.time()

        print('%r (...) %2.2f sec' %
              (method.__name__, te - ts))
        return result

    return timed


class DictWrapper:
    def __init__(self, _dict):
        self.dict = _dict

    def __getattr__(self, entry):
        rtn = self.dict[entry]
        if isinstance(rtn, dict):
            return DictWrapper(rtn)
        else:
            return rtn

    # def __setattr__(self, entry, val):
    #     self.dict[entry] = val

    def __setitem__(self, entry, val):
        self.dict[entry] = val

    def __getitem__(self, entry):

        return self.__getattr__(entry)


travis/install_cuda.sh
 # install repo packages
CUDA_REPO_PKG=cuda-repo-ubuntu1404_7.5-18_amd64.deb
wget http://developer.download.nvidia.com/compute/cuda/repos/ubuntu1404/x86_64/$CUDA_REPO_PKG
dpkg -i $CUDA_REPO_PKG
rm $CUDA_REPO_PKG

if $WITH_CUDNN ; then
ML_REPO_PKG=nvidia-machine-learning-repo-ubuntu1404_4.0-2_amd64.deb
wget http://developer.download.nvidia.com/compute/machine-learning/repos/ubuntu1404/x86_64/$ML_REPO_PKG
dpkg -i $ML_REPO_PKG
fi

# update package lists
apt-get -y update

# install packages
CUDA_PKG_VERSION="7-5"
CUDA_VERSION="7.5"
apt-get install -y --no-install-recommends \
cuda-core-$CUDA_PKG_VERSION \
cuda-cudart-dev-$CUDA_PKG_VERSION \
cuda-cublas-dev-$CUDA_PKG_VERSION \
cuda-curand-dev-$CUDA_PKG_VERSION
# manually create CUDA symlink
ln -s /usr/local/cuda-$CUDA_VERSION /usr/local/cuda

if $WITH_CUDNN ; then
	apt-get install -y --no-install-recommends libcudnn7-dev
fi


farneback3d/skeleton.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""
This is a skeleton file that can serve as a starting point for a Python
console script. To run this script uncomment the following line in the
entry_points section in setup.cfg:

    console_scripts =
     fibonacci = farneback3d.skeleton:run

Then run `python setup.py install` which will install the command `fibonacci`
inside your current environment.
Besides console scripts, the header (i.e. until _logger...) of this file can
also be used as template for Python modules.

Note: This skeleton file can be safely removed if not needed!
"""
from __future__ import division, print_function, absolute_import

import argparse
import sys
import logging

from farneback3d import __version__

__author__ = "Stephan Seitz"
__copyright__ = "Stephan Seitz"
__license__ = "none"

_logger = logging.getLogger(__name__)


def fib(n):
    """Fibonacci example function

    Args:
      n (int): integer

    Returns:
      int: n-th Fibonacci number
    """
    assert n > 0
    a, b = 1, 1
    for i in range(n-1):
        a, b = b, a+b
    return a


def parse_args(args):
    """Parse command line parameters

    Args:
      args ([str]): command line parameters as list of strings

    Returns:
      :obj:`argparse.Namespace`: command line parameters namespace
    """
    parser = argparse.ArgumentParser(
        description="Just a Fibonnaci demonstration")
    parser.add_argument(
        '--version',
        action='version',
        version='farneback3d {ver}'.format(ver=__version__))
    parser.add_argument(
        dest="n",
        help="n-th Fibonacci number",
        type=int,
        metavar="INT")
    parser.add_argument(
        '-v',
        '--verbose',
        dest="loglevel",
        help="set loglevel to INFO",
        action='store_const',
        const=logging.INFO)
    parser.add_argument(
        '-vv',
        '--very-verbose',
        dest="loglevel",
        help="set loglevel to DEBUG",
        action='store_const',
        const=logging.DEBUG)
    return parser.parse_args(args)


def setup_logging(loglevel):
    """Setup basic logging

    Args:
      loglevel (int): minimum loglevel for emitting messages
    """
    logformat = "[%(asctime)s] %(levelname)s:%(name)s:%(message)s"
    logging.basicConfig(level=loglevel, stream=sys.stdout,
                        format=logformat, datefmt="%Y-%m-%d %H:%M:%S")


def main(args):
    """Main entry point allowing external calls

    Args:
      args ([str]): command line parameter list
    """
    args = parse_args(args)
    setup_logging(args.loglevel)
    _logger.debug("Starting crazy calculations...")
    print("The {}-th Fibonacci number is {}".format(args.n, fib(args.n)))
    _logger.info("Script ends here")


def run():
    """Entry point for console_scripts
    """
    main(sys.argv[1:])


if __name__ == "__main__":
    run()


farneback3d/_filtering.py
import os
import numpy as np
import pycuda.autoinit
import pycuda.driver as cuda
import pycuda.gpuarray as gpuarray
from pycuda.compiler import SourceModule

from farneback3d._utils import divup

with open(os.path.join(os.path.dirname(__file__), 'filtering.cu')) as f:
    read_data = f.read()
f.closed

mod = SourceModule(read_data)


def _call_smooth_cuda_gauss(gpuimg, gpuresult, sigma, kernelsize,  filter_mask=None):

    filter_nonzeros_only = filter_mask is not None

    if len(gpuimg.shape) == 2:
        block = (32, 32, 1)
        grid = (int(divup(gpuimg.shape[1], block[0])),
                int(divup(gpuimg.shape[0], block[1])), 1)

        assert not filter_nonzeros_only, "not implemented yet for 2d"
        assert filter_mask is None

        convolve2d = mod.get_function("convolve2d_gauss")
        convolve2d(gpuimg,
                   gpuresult,
                   np.int32(gpuimg.shape[1]),
                   np.int32(gpuimg.shape[0]),
                   np.int32(kernelsize),
                   np.int32(kernelsize),
                   np.float32(sigma * sigma),
                   np.float32(sigma * sigma),
                   grid=grid, block=block)
    elif len(gpuimg.shape) == 3:
        block = (32, 32, 1)
        grid = (int(divup(gpuimg.shape[2], block[0])),
                int(divup(gpuimg.shape[1], block[1])), 1)

        if filter_mask is None:
            convolve3d = mod.get_function("convolve3d_gauss")
            convolve3d(gpuimg,
                       gpuresult,
                       np.int32(gpuimg.shape[2]),
                       np.int32(gpuimg.shape[1]),
                       np.int32(gpuimg.shape[0]),
                       np.int32(kernelsize),
                       np.int32(kernelsize),
                       np.int32(kernelsize),
                       np.float32(sigma * sigma),
                       np.float32(sigma * sigma),
                       np.float32(sigma * sigma),
                       np.int32(filter_nonzeros_only),
                       grid=grid, block=block)
        else:
            convolve3d = mod.get_function("convolve3d_gauss_with_mask")
            convolve3d(gpuimg,
                       gpuresult,
                       filter_mask,
                       np.int32(gpuimg.shape[2]),
                       np.int32(gpuimg.shape[1]),
                       np.int32(gpuimg.shape[0]),
                       np.int32(kernelsize),
                       np.int32(kernelsize),
                       np.int32(kernelsize),
                       np.float32(sigma * sigma),
                       np.float32(sigma * sigma),
                       np.float32(sigma * sigma),
                       np.int32(filter_nonzeros_only),
                       grid=grid, block=block)
    else:
        raise(ValueError('only dimensions 2 or 3 allowed'))

    return gpuresult


def smooth_cuda_gauss(img, sigma, kernelsize, rtn_gpu=None, filter_mask=None):
    if not rtn_gpu:
        rtn_gpu = gpuarray.GPUArray(img.shape, np.float32)

    if isinstance(img, np.ndarray):

        img = np.ascontiguousarray(img, np.float32)

        img_gpu = gpuarray.to_gpu(img)

        _call_smooth_cuda_gauss(img_gpu, rtn_gpu, sigma,
                                kernelsize, filter_mask)

        return rtn_gpu.get()
    elif isinstance(img, gpuarray.GPUArray):
        # TODO inplace algorithm
        _call_smooth_cuda_gauss(img, rtn_gpu,
                                sigma,  kernelsize, filter_mask)

    return rtn_gpu


farneback3d/_farneback3d.py
"""
3d Python implementation of 2d optical flow "opencv/modules/video/src/optflowgf.cpp"

"""
# TODO make class save fft kernels for fast filtering

import numpy as np
import scipy.ndimage as sciimg
import pycuda.driver as cuda
import pycuda.gpuarray as gpuarray
import pycuda.autoinit
from pycuda.compiler import SourceModule
import os
import json

import farneback3d._utils
from farneback3d._utils import divup
import farneback3d._filtering

_NUM_POLY_COEFFICIENTS = 10
_MIN_VOL_SIZE = 32


class Farneback:

    def __init__(self,
                 pyr_scale=0.9,
                 levels=15,
                 winsize=9,
                 num_iterations=5,
                 poly_n=5,
                 poly_sigma=1.2,
                 use_gaussian_kernel:  bool= True,
                 use_initial_flow=None,
                 quit_at_level=None,
                 use_gpu=True,
                 upscale_on_termination=True,
                 fast_gpu_scaling=True,
                 vesselmask_gpu=None
                 ):
        self.pyr_scale = pyr_scale
        self.levels = levels
        self.winsize = winsize
        self.num_iterations = num_iterations
        self.poly_n = poly_n
        self.poly_sigma = poly_sigma
        self.use_gaussian_kernel = use_gaussian_kernel
        self.use_initial_flow = use_initial_flow
        self.use_gpu = use_gpu
        self.upscale_on_termination = upscale_on_termination
        self._fast_gpu_scaling = fast_gpu_scaling
        self.quit_at_level = quit_at_level
        self._dump_everything = False
        self._show_everything = False
        self._vesselmask_gpu = vesselmask_gpu
        self._resize_kernel_size_factor = 4
        self._max_resize_kernel_size = 9

        with open(os.path.join(os.path.dirname(__file__), 'farneback_kernels.cu')) as f:
            read_data = f.read()
        f.closed

        mod = SourceModule(read_data)
        self._update_matrices_kernel = mod.get_function(
            'FarnebackUpdateMatrices')
        self._invG_gpu = mod.get_global('invG')[0]
        self._weights_gpu = mod.get_global('weights')[0]
        self._poly_expansion_kernel = mod.get_function('calcPolyCoeficients')
        self._warp_kernel = mod.get_function('warpByFlowField')
        self._r1_texture = mod.get_texref('sourceTex')
        self._solve_equations_kernel = mod.get_function('solveEquationsCramer')

    def load_parameters_from_config_file(self, config_file=None):
        if not config_file:
            config_file = os.path.join(
                os.path.dirname(__file__), "optflow.json")

        with open(config_file) as file:
            config_str = file.read()

        config = farneback3d._utils.DictWrapper(json.loads(config_str))

        self.levels = config.levels
        self.num_iterations = config.num_iterations
        self.poly_n = config.poly_n
        self.quit_at_level = config.quit_at_level
        self.upscale_on_termination = config.upscale_on_termination
        self.poly_sigma = config.poly_sigma
        self.winsize = config.winsize
        self.use_gaussian_kernel = config.use_gaussian_kernel
        self._resize_kernel_size_factor = config.resize_kernel_size_factor
        self._max_resize_kernel_size = config.max_resize_kernel_size

    def write_parameters_from_config_file(self, config_file):
        config_dict = {}
        config_dict["levels"] = self.levels
        config_dict["num_iterations"] = self.num_iterations
        config_dict["poly_n"] = self.poly_n
        config_dict["quit_at_level"] = self.quit_at_level
        config_dict["upscale_on_termination"] = self.upscale_on_termination
        config_dict["poly_sigma"] = self.poly_sigma
        config_dict["winsize"] = self.winsize
        config_dict["use_gaussian_kernel"] = self.use_gaussian_kernel
        config_dict["resize_kernel_size_factor"] = self._resize_kernel_size_factor
        config_dict["max_resize_kernel_size"] = self._max_resize_kernel_size
        with open(config_file, 'w') as file:
            json.dump(config_dict, file)

    def calc_flow(self,
                  cur_vol: np.ndarray,
                  next_vol: np.ndarray
                  ):

        dim = len(cur_vol.shape)

        assert dim == 3, 'wrong dimension'
        assert self.quit_at_level < self.levels
        assert cur_vol.shape == next_vol.shape
        assert cur_vol.dtype == np.float32, 'wrong dtype'
        assert next_vol.dtype == np.float32, 'wrong dtype'
        assert not self.use_initial_flow

        prev_flow_gpu = None
        imgs = [gpuarray.to_gpu(next_vol), gpuarray.to_gpu(cur_vol)]
        flow_gpu = None

        for k in range(self.levels, -1, -1):
            print('Scale %i' % k)

            if k == self.quit_at_level:
                if k != -1 and self.upscale_on_termination and prev_flow_gpu is not None:
                    flow_gpu = gpuarray.GPUArray(
                        [dim, *cur_vol.shape], cur_vol.dtype)

                    prev_flow_gpu *= 1 / scale
                    for i in range(3):
                        self._resize(
                            prev_flow_gpu[i], flow_gpu[i], dst_shape=cur_vol.shape)

                break

            scale = self.pyr_scale**k

            if np.any(np.array(cur_vol.shape) * scale < _MIN_VOL_SIZE):
                continue

            sigma = (1. / scale - 1) * 0.5
            smooth_sz = int(sigma * self._resize_kernel_size_factor + 0.5) | 1
            smooth_sz = max(smooth_sz, 3)
            smooth_sz = min(smooth_sz, self._max_resize_kernel_size)

            scale_shape = [int(np.round(x * scale)) for x in cur_vol.shape]

            if prev_flow_gpu is not None:
                flow_gpu *= 1 / self.pyr_scale
                scaled_flow_gpu = gpuarray.GPUArray(
                    [3, *scale_shape], np.float32)
                for i in range(3):
                    self._resize(
                        flow_gpu[i], scaled_flow_gpu[i], dst_shape=scale_shape)

                flow_gpu = scaled_flow_gpu

            else:
                flow_gpu = gpuarray.zeros([dim, *scale_shape], np.float32)

            R = gpuarray.GPUArray(
                [2, _NUM_POLY_COEFFICIENTS - 1, *flow_gpu.shape[1:]], np.float32)
            M = gpuarray.GPUArray([9, *flow_gpu.shape[1:]], np.float32)

            for i in range(2):
                if k == 0:
                    I = imgs[i]
                else:
                    I = farneback3d._filtering.smooth_cuda_gauss(imgs[i], sigma,
                                                                 smooth_sz)
                    I = self._resize(I, scaling=scale)

                self._FarnebackPolyExp(I, R[i], self.poly_n, self.poly_sigma)

            # dsareco.visualization.imshow(R[0])
            # dsareco.visualization.imshow(R[1])

            self._FarnebackUpdateMatrices_gpu(R[0], R[1], flow_gpu, M)

            # dsareco.visualization.imshow(M,"M")

            for i in range(self.num_iterations):
                print('iteration %i' % i)
                if self.use_gaussian_kernel:
                    self._FarnebackUpdateFlow_GaussianBlur_gpu(
                        R[0], R[1], flow_gpu, M, self.winsize, i < self.num_iterations - 1)
                else:
                    raise ValueError('only use_gaussian_kernel implemented')
                # dsareco.utils.imshow(np.moveaxis(flow,3,0),"%i, scale %i" %(i,k))
                # dsareco.vtk.imageToVTK('/dos/d/output/flow_%03d_%03d' %(k,i), pointData={'x' : np.ascontiguousarray(flow[:,:,:,0]),  'y': np.ascontiguousarray(flow[:,:,:,1]), 'z': np.ascontiguousarray(flow[:,:,:,2])})

            prev_flow_gpu = flow_gpu

        if flow_gpu is not None:
            return flow_gpu.get()
        else:
            return np.zeros([3, *cur_vol.shape], np.float32)

    def _FarnebackPrepareGaussian(self, n, sigma):
        if sigma < 1e-7:
            sigma = n * 0.3

        x = np.arange(-n, n + 1, dtype=np.float32)
        g = np.exp(-(x * x) / (2 * sigma * sigma))
        g /= np.sum(g)
        # xg = x * g
        # xxg = x * xg

        G = np.zeros(
            (_NUM_POLY_COEFFICIENTS, _NUM_POLY_COEFFICIENTS), np.float32)

        G_half = np.zeros((10, 2 * n + 1, 2 * n + 1, 2 * n + 1), np.float32)

        # G:  sum_xyz weight_xyz *(1 x y z xx yy zz xy xz yz) * (1 x y z xx yy zz xy xz yz)^T

        for z in range(-n, n + 1):
            for y in range(-n, n + 1):
                for x in range(-n, n + 1):
                    gauss_weight = g[z + n] * g[y + n] * g[x + n]
                    base_vector = np.atleast_2d(np.array(
                        [1, x, y, z, x * x, y * y, z * z, x * y, x * z, y * z], np.float32))
                    matrix = np.matmul(np.transpose(base_vector), base_vector)
                    G += matrix * gauss_weight
                    G_half[:, z + n, y + n, x + n] = gauss_weight * base_vector

        invG = np.linalg.inv(G)

        return invG, G_half

    def _FarnebackPolyExp(self, img_gpu, poly_coefficients_gpu, n, sigma):
        assert img_gpu.size == poly_coefficients_gpu[0].size
        assert img_gpu.dtype == np.float32
        assert poly_coefficients_gpu.dtype == np.float32

        invG, G_half = self._FarnebackPrepareGaussian(n, sigma)

        block = (32, 32, 1)
        grid = (int(divup(img_gpu.shape[2], block[0])),
                int(divup(img_gpu.shape[1], block[1])), 1)

        cuda.memcpy_htod(self._invG_gpu, invG)
        cuda.memcpy_htod(self._weights_gpu, G_half)

        self._poly_expansion_kernel(img_gpu,
                                    poly_coefficients_gpu,
                                    np.int32(img_gpu.shape[2]),
                                    np.int32(img_gpu.shape[1]),
                                    np.int32(img_gpu.shape[0]),
                                    np.int32(2 * n + 1),
                                    grid=grid, block=block)

    def _FarnebackUpdateMatrices_gpu(self, R0_gpu, R1_gpu, flow_gpu, M_gpu):

        R1_warped_gpu = gpuarray.empty_like(R1_gpu)

        block = (32, 32, 1)
        grid = (int(divup(flow_gpu.shape[3], block[0])),
                int(divup(flow_gpu.shape[2], block[1])), 1)

        for i in range(_NUM_POLY_COEFFICIENTS - 1):
            farneback3d._utils.ndarray_to_float_tex(
                self._r1_texture, R1_gpu[i])
            self._warp_kernel(
                flow_gpu,
                R1_warped_gpu[i],
                np.int32(flow_gpu.shape[3]),
                np.int32(flow_gpu.shape[2]),
                np.int32(flow_gpu.shape[1]),
                np.float32(1),
                np.float32(1),
                np.float32(1),
                block=block, grid=grid)

        self._update_matrices_kernel(R0_gpu,
                                     R1_warped_gpu,
                                     flow_gpu,
                                     M_gpu,
                                     np.int32(flow_gpu.shape[3]),
                                     np.int32(flow_gpu.shape[2]),
                                     np.int32(flow_gpu.shape[1]),
                                     block=block, grid=grid)

    def _FarnebackUpdateFlow_GaussianBlur_gpu(self, poly_coefficients0, poly_coefficients1, flow_gpu, M, winsize, update_matrices):
        sigma = self.winsize * 0.3

        M_filtered_gpu = gpuarray.GPUArray(M.shape, M.dtype)

        for i in range(M.shape[0]):
            farneback3d._filtering.smooth_cuda_gauss(
                M[i], sigma, winsize, rtn_gpu=M_filtered_gpu[i])

        block = (32, 32, 1)
        grid = (int(divup(flow_gpu.shape[3], block[0])),
                int(divup(flow_gpu.shape[2], block[1])), 1)

        self._solve_equations_kernel(M_filtered_gpu, flow_gpu, np.int32(flow_gpu.shape[3]), np.int32(
            flow_gpu.shape[2]), np.int32(flow_gpu.shape[1]), block=block, grid=grid)

        if update_matrices:
            self._FarnebackUpdateMatrices_gpu(
                poly_coefficients0, poly_coefficients1, flow_gpu, M)

    def _resize(self, src_vol, dst_vol=None, dst_shape=None, scaling=None):

        if self._fast_gpu_scaling:
            return farneback3d._utils.resize_gpu(src_vol, dst_vol, dst_shape, scaling)
        else:
            if scaling:
                scaling_vec = [scaling, scaling, scaling]
            else:
                assert dst_shape is not None
                scaling_vec = [dst_shape[i] / src_vol.shape[i]
                               for i in range(len(dst_shape))]

            result = sciimg.zoom(src_vol.get(), scaling_vec)

            if dst_vol is None:
                dst_vol = gpuarray.to_gpu(result)
            else:
                dst_vol.set(result)

            return dst_vol


def warp_by_flow(vol, flow3d):
    """
    Only used for testing at the moment
    (warps currently backward)

    """
    with open(os.path.join(os.path.dirname(__file__), 'farneback_kernels.cu')) as f:
        read_data = f.read()
    f.closed

    mod = SourceModule(read_data)
    interpolation_kernel = mod.get_function('warpByFlowField')

    r1_texture = mod.get_texref('sourceTex')

    farneback3d._utils.ndarray_to_float_tex(r1_texture, vol)
    rtn_gpu = gpuarray.GPUArray(vol.shape, vol.dtype)
    flow_gpu = gpuarray.to_gpu(flow3d)

    block = (32, 32, 1)
    grid = (int(divup(flow3d.shape[3], block[0])),
            int(divup(flow3d.shape[2], block[1])), 1)

    interpolation_kernel(
        flow_gpu,
        rtn_gpu,
        np.int32(flow_gpu.shape[3]),
        np.int32(flow_gpu.shape[2]),
        np.int32(flow_gpu.shape[1]),
        np.float32(1),
        np.float32(1),
        np.float32(1),
        block=block, grid=grid)

    return rtn_gpu.get()
